components/movementsensor/rtkutils/i2c_data_reader.go

```diff
@@ -0,0 +1,215 @@
+}
+
+func (dr *PmtkI2cDataReader) readData() ([]byte, error) {
+	handle, err := dr.bus.OpenHandle(dr.addr)
+	if err != nil {
+		dr.logger.CErrorf(dr.cancelCtx, "can't open gps i2c %s", err)
+		return nil, err
+	}
+	defer utils.UncheckedErrorFunc(handle.Close)
+
+	buffer, err := handle.Read(dr.cancelCtx, 1024)
+	if err != nil {
+		dr.logger.CErrorf(dr.cancelCtx, "failed to read handle %s", err)
+		return nil, err
+	}
+
+	return buffer, nil
+}
+
+// start spins up a background coroutine to read data from the I2C bus and put it into the channel
+// of complete messages.
+func (dr *PmtkI2cDataReader) start() {
+	dr.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer dr.activeBackgroundWorkers.Done()
+		defer close(dr.data)
+
+		strBuf := ""
+		for {
+			select {
+			case <-dr.cancelCtx.Done():
+				return
+			default:
+			}
+
+			buffer, err := dr.readData()
+			if err != nil {
+				dr.logger.CErrorf(dr.cancelCtx, "failed to read data, retrying: %s", err)
+				continue
+			}
+
+			for _, b := range buffer {
+				// PMTK uses CRLF line endings to terminate sentences, but just LF to blank data.
+				// Since CR should never appear except at the end of our sentence, we use that to
+				// determine sentence end. LF is merely ignored.
+				if b == 0x0D { // 0x0D is the ASCII value for a carriage return
+					if strBuf != "" {
+						// Sometimes we miss "$" on the first message of the buffer. If the first
+						// character we read is a "G", it's likely that this has occurred, and we
+						// should add a "$" at the beginning.
+						if strBuf[0] == 0x47 { // 0x47 is the ASCII value for "G"
+							strBuf = "$" + strBuf
+						}
+
+						dr.data <- strBuf
+						strBuf = ""
+
+						// Check if we're supposed to shut down again. Perhaps we waited a long
+						// time to put data into the channel.
+						select {
+						case <-dr.cancelCtx.Done():
+							return
+						default:
+						}
+					}
+				} else if b != 0x0A && b < 0x7F { // only append valid (printable) bytes
+					strBuf += string(b)
+				}
+			}
+		}
+	})
+}
+
+// Messages returns the channel of complete NMEA sentences we have read off of the device. It's part
+// of the DataReader interface.
+func (dr *PmtkI2cDataReader) Messages() chan string {
+	return dr.data
+}
+
+// Close is part of the DataReader interface. It shuts everything down.
+func (dr *PmtkI2cDataReader) Close() error {
+	dr.cancelFunc()
+	// If the background coroutine is trying to put a new line of data into the channel, it won't
+	// notice that we've canceled it until something tries taking the line out of the channel. So,
+	// let's try to read that out so the coroutine isn't stuck. If the background coroutine shut
+	// itself down already, the channel will be closed and reading something out of it will just
+	// return the empty string.
+	<-dr.data
+	dr.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+// PMTK checksums commands by XORing together each byte.
+func addChk(data []byte) []byte {
```

Comment:
We should get rid of `addChk `and `checksum` function and use `PMTKAddChk` and `PMTKChecksum` from `movementsensor/utils.go`. They are duplicates.

Reply:
Thank you for noticing!

Reply:
take another look